### DISCUSSION TITLE
"mode selection" at boot

### DISCUSSION THREAD

**Discussion by gringow831** [tttapa/Control-Surface#856] [Q&A]
I have been trying to save the "selected mode" at boot so next time the controller loads the last setting is already selected rather than selecting every time but not much luck.

The theory is that when the arduino is booting, if there is a button pressed, a mode or another will be selected for the session. (not when it's already loaded). If no button is pressed at boot, continue with the previous settings (mode). So enable and disable banks. In this test, just NoteButton for now. There will be more than 2 modes but 2 for now in case I can make it work
Any help would be appreciated.

```ruby
#include <Control_Surface.h>
#include <EEPROM.h>

USBMIDI_Interface midi;

 NoteButton first2[] = {
  {5, MCU::SELECT_1},
  {7, MCU::SELECT_2},
 };
 
 NoteButton second2[] = {
  {5, MCU::SELECT_3},
  {7, MCU::SELECT_4},
};

Button btn5 {5}; 
Button btn7 {7};

uint8_t savedSetting; //for the first2
uint8_t savedSetting_1; //for the second2

void setup() {
  savedSetting = EEPROM.read(0);
  if (savedSetting != first2[2].isEnabled())
  {    }
  else 
    savedSetting = first2[2].isEnabled();


  savedSetting_1 = EEPROM.read(1);
  if (savedSetting_1 != second2[2].isEnabled())
  {    }
  else 
    savedSetting_1 = second2[2].isEnabled();
  
NoteButton::disable(second2); //disable the second2 NoteButton

//if pressed while boot enable the first3 button (IT BREAKS THE CODE IF THE FIRST2 ARE ALREADY SELECTED BUT NECESARY IF SAVED SECOND2 TO GO BACK)
if (btn5.update()) {
  if (btn5.getState()==Button::Released){
    }
  if (btn5.getState()==Button::Falling){
  NoteButton::enable(first2);
  NoteButton::disable(second2);
    }}

//enable second3 and disable first3 NoteButtons
if (btn7.update()) {
    if (btn7.getState()==Button::Released){
    }
   if (btn7.getState()==Button::Falling){
  NoteButton::disable(first2);
  NoteButton::enable(second2);
     }
}
}

void loop() {
  Control_Surface.loop();

  auto currSetting = first2[2].isEnabled();
  if (savedSetting != currSetting) {
    EEPROM.write(0, currSetting);
    savedSetting = currSetting;
  }


  auto currSetting_1 = second2[2].isEnabled();
  if (savedSetting_1 != currSetting_1) {
    EEPROM.write(1, currSetting_1);
    savedSetting_1 = currSetting_1;
  }
  }
```

**Comment by tttapa** [comment]
`first2[2].isEnabled()` will always evaluate to true in the setup, because it cannot have been disabled yet (all MIDI elements start off as enabled, you have to disable them explicitly if you don't want that).

There's no real reason to use the Button class for reading the state of the pins, you can just use `digitalRead()` (don't forget to call `pinMode()` first).

Having two NoteButtons on the same pin is usually a bad idea. Are you sure you don't want to use the Bank class instead? https://tttapa.github.io/Control-Surface-doc/Doxygen/da/dc1/FAQ.html#faq-change-address-runtime

The call to `Control_Surface.begin()` seems to be missing.

For project-specific questions, you might get more help on <https://forum.arduino.cc>.

**Reply by gringow831** [reply]
Thank you for that. 
I did try digitalread and pinMode and still don’t seem to find the way to do it. I looked at the tutorial off https://randomnerdtutorials.com/arduino-eeprom-explained-remember-last-led-state/
To try replicating  this with control surface. 

In the example I am planning to do there are many groups and all work within each “mode” which have managed to made it work with a dipswitch. This selects every “mode” but I would like to get rid of the switch and just use the buttons that I am already planning. 

with long press I managed to changed the “modes” while on loop but not remember the last mode.  Is there a way to do that?

Is it possible to actually do it during the setup if the button is press while boot?

I’m guessing if everything is disabled at the beginning might be complicated since I first would need to select a mode before it actually works on anything… 

How do I save the mode done by the  long press as this is disable/enable rather than banks. 
All banks are already saved whenever each one is selected and working
